### ISSUE TITLE
Request: allow specifying cache size as percentage (vs. size-based `cache_size_mb`)

### ISSUE THREAD

**Issue by rrjjvv** [earthly/earthly#1723]
I'm overloading this request with some other feedback/questions which might warrant their own issues, but figured I'd start simple.  But the fundamental request is as stated: it would be great to have something like `cache_size_pct` as a companion to `cache_size_mb`.  Or overload `cache_size_mb` to accept a percentage, if that was preferred.

My own machines, as well as build machines I manage, are of different sizes.  I would prefer to have one static config file with a static percentage (which adapts to the host at startup).  Dynamically generated config files (with calculated sizes) aren't impossible, but isn't ideal.

Background
---
I had a build target that was taking a very long time to execute on back-to-back executions, when theoretically everything should have been cached.  It seemed to also be the same sub-target that was building from scratch, and for the life of me, I could not figure out what was causing a cache bust.  It then occurred to me it might not be a cache bust per-se, but a cache expiration/eviction.  And it was.

I was always aware of the `cache_size_mb` setting, but the language says
> If you would like to limit the cache size more aggressively, you can specify a different limit 

To me, that implied that the default value was often too large for people, so it hadn't originally occurred to me that the opposite could occur (where the default is too small).  After some experimentation and strolling through source code, I found that (on linux) the default is [10% of the filesystem containing the cache](https://github.com/moby/buildkit/blob/master/cmd/buildkitd/config/gcpolicy_unix.go#L16).  If my laptop had everything on the root filesystem (which might be the common case with end-users), I'd have a healthy 50GB cache.  But `/var` (thus docker) is on its own filesystem, leaving me with a puny 4GB cache.  The smaller targets all fit in cache just fine, but the larger aggregate targets were exceeding that, and very quickly.  After completion, the target uses 8GB of cache, but grows up to 10GB.  After specifying a larger size, my problematic target went from about 30 minutes down to 6 minutes.

Specifying a custom setting results in [your template](https://github.com/earthly/earthly/blob/main/buildkitd/entrypoint.sh#L136-L138) getting rendered, which results in a policy like
```
GC Policy rule#0:
        All:            false
        Filters:        type==source.local type==source.git.checkout
        Keep Bytes:     1.5GB
GC Policy rule#1:
        All:            true
        Keep Bytes:     15GB
```
but the default configuration (which I believe is a result of [this](https://github.com/moby/buildkit/blob/master/cmd/buildkitd/config/gcpolicy.go#L11-L29)) results in a policy that differs in ways beyond just size,
```
GC Policy rule#0:
        All:            false
        Filters:        type==source.local,type==exec.cachemount,type==source.git.checkout
        Keep Duration:  48h0m0s
        Keep Bytes:     512MB
GC Policy rule#1:
        All:            false
        Keep Duration:  1440h0m0s
        Keep Bytes:     4GB
GC Policy rule#2:
        All:            false
        Keep Bytes:     4GB
GC Policy rule#3:
        All:            true
        Keep Bytes:     4GB
```
Is that intentional?  That's fine if it is, but to a layman like me, I'd think that if your overridden policy is "better" for earthly usage, then you'd apply it even if the user allows buildkit to choose the size values.  The one scenario where the deviation might make sense is if your policy is better for low-space scenarios (the implied primary use case of customizing `cache_size_mb`).  Regardless, this doesn't have any direct bearing on my request.  I just noticed it during my journey and wanted to bring it up in case it was not known.

Suggestions/Feedback
---
1. In the [config doc](https://docs.earthly.dev/docs/earthly-config#cache_size_mb), call out that the adaptive piece is for "how much space is available to docker", not "how much space is available on your system".  Most docker users probably understand that, but for users that have dedicated partitions for docker, that verbiage change would help indicate how significant that setting is.
2. In the [related page](https://docs.earthly.dev/docs/guides/cache#specifying-cache-size), reword it in a way that indicates that increasing the value may also be desirable/needed.  Or maybe even call this scenario explicitly.
3. In the `earthly-buildkitd` startup logs, output the cache size that will be used if not specified.

I'm not sure the first two would have prevented my problem, but I do know that exploring cache size would have been on my list of things to look at and not a last resort.  The third would have cut down on diagnostic time.  It wasn't until I put a `watch` on the `earthly-cache` volume's size (and saw it creep to 4GB and then stop) that I realized cache size was my issue.  (Tracking down where that 4GB came from is another story, but I now know that's entirely a buildkit issue, not earthly.)

---

Let me know if my original request is legitimate.  If so, it might be something I'd be capable of doing myself (as well as my third suggestion).

**Comment by vladaionescu** [comment]
Sounds good to me

**Comment by rrjjvv** [comment]
Would you prefer a dedicated setting, or overload?

Assuming a new setting:
1. Do you have a name in mind?
2. Any preference on the value's (logical) type?
3. How should this new setting and `cache_size_mb` interact?

If you have no opinions, I'd probably go with `cache_size_pct`, have it be integral (`0 < x <= 100`, with `0` denoting "use defaults" like `cache_size_mb` does), and have them be mutually exclusive (solely to keep it simple).

If you want an overload, there's just one question: how to disambiguate the two usages (the obvious choices being `.6` or `60%`).

**Comment by vladaionescu** [comment]
Yeah, agreed, I wouldn't overload mainly because of the current name (`cache_size_mb` suggests this should be a number of megabytes). I think a separate percentage setting makes sense.

Another idea is to use the minimum of the two values. So if both are set, and `cache_size_pct` is set to `15`(which e.g. results in `150mb`) and `cache_size_mb` is set to `200`, then the limit is taken as `150mb`, because that's the minimum of the two.

**Comment by rrjjvv** [comment]
@vladaionescu I have this implementation done, though the code ended up being very different then what I thought was needed.  As a direct result, the third item from my wishlist (outputting the calculated size if not provided) wasn't possible and the tests are minimal (some rewrites, breaking from current conventions, or tricks I'm not privy too would be required to do anything substantial).

The existing (and new) tests all pass except for ones specific to podman. but they also fail (and in the same way) in a pristine checkout (using both the release and built binaries), so unrelated to my changes.

To get those tests running (consistently) required a number of unrelated changes to remove the reliance on docker.io (in favor of a proxy registry).  This was a relatively large change; most is trivial boilerplate, but a small part is relatively intrusive (for a first-time contributor).  I also have totally unrelated changes (typos, broken links, etc. as I encountered them).

For the trivial, non-related changes, would you like them broken out to a separate PR, or bundled in with my implementation?

And are you interested in my registry-related changes?  It's probably "too big" of a change to tack on to the original work, but I'd be happy to do so.  Do you want a separate PR for this, and if so, do you want/need a dedicated issue filed for it?  I'd be happy to elaborate offline as well.

**Comment by vladaionescu** [comment]
Ideally small, isolated PRs are most useful, but small fixes along the way should be fine to tack in the same PR. Draft PRs that are a bit bigger are also ok for getting feedback, if you need it.

If something gets more involved (like it sounds it's getting with the podman/registry-related things), it's best to put that in a separate PR. I know that @dchw has been working on some registry-related podman fixes. He's probably the best person to chat with - you can grab him over Slack as `Corey Larson`.

**Comment by rrjjvv** [comment]
Regarding podman issues, I'm guessing #1682 is probably the fix (the description, as well as conversation in the linked issue, definitely line up).  Incidentally, the change related to cgroups today (#1703) are breaking a large percentage of existing tests on my host (which does use v2), but trying to isolate any commonalities for you folks.

For clarity, my registry-related (proposed) changes have nothing to do with podman itself; it'll make more sense when I create a draft PR.  (I was hitting rate-limits with hub.  The `buildkit-additional-args` goes a long way, but the buildkit-in-buildkit as well as d-in-d were still problems.  I 'finished' the buildkit-in-buildkit piece that you folks started, and got the d-in-d wrapper to use a mirror as well.  I don't know if I've *totally* stopped any hits to hub, but certainly enough to allow me to run all the tests with no issue.)

I should get some stuff pushed for review in a couple hours.

**Comment by vladaionescu** [comment]
CC @alexcb - FYI since you have been reviewing the cache pct PR.

**Comment by rrjjvv** [comment]
Couple notes/updates:

1. In my original request/report, I mentioned a discrepancy in the GC configuration when template-generated (`cache_size_mb` or `cache_size_pct`) vs. the buildkitd default.  I don't think anyone mentioned whether that was by design or not.  Even if not, there didn't appear to be a way to derive from the buildkitd defaults in a future-proof way.  For those two reasons, I didn't bother touching it.  I'd be happy to change the template to match the default if desired, though it could get out of sync in the future.
2. I alluded to changes in allowing buildkitd-in-buildkitd to use registry (and  insecure) mirrors for tests; I did create a separate PR for that
3. I also mentioned similar changes regarding the embedded docker (to allow _it_ to use mirrors as well).  I did not push that code since I had reason to think it was working "on accident" and not actually needed, but I now know it truly was (and is required to be truly dockerhub-independent).  I'm not pushing that yet though since 1) if you're not receptive to the type of changes in my b-in-b PR, you probably won't to my b-in-d changes either, and 2) I know it can be made a little better.
